app/src/main/java/io/homeassistant/companion/android/util/TLSWebViewClient.kt

```diff
@@ -0,0 +1,106 @@
+package io.homeassistant.companion.android.util
+
+import android.app.Activity
+import android.content.Context
+import android.content.ContextWrapper
+import android.os.Build
+import android.security.KeyChain
+import android.security.KeyChainAliasCallback
+import android.webkit.ClientCertRequest
+import android.webkit.WebView
+import android.webkit.WebViewClient
+import androidx.annotation.RequiresApi
+import androidx.appcompat.app.AppCompatActivity
+import androidx.lifecycle.lifecycleScope
+import io.homeassistant.companion.android.common.data.keychain.KeyChainRepository
+import kotlinx.coroutines.launch
+import java.security.Principal
+import java.security.PrivateKey
+import java.security.cert.X509Certificate
+import javax.inject.Inject
+
+open class TLSWebViewClient @Inject constructor(private var keyChainRepository: KeyChainRepository) : WebViewClient() {
+
+    var isTLSClientAuthNeeded = false
+        private set
+
+    var hasUserDeniedAccess = false
+        private set
+
+    private var key: PrivateKey? = null
+    private var chain: Array<X509Certificate>? = null
+
+    private fun getActivity(context: Context?): Activity? {
+        if (context == null) {
+            return null
+        } else if (context is ContextWrapper) {
+            return if (context is Activity) {
+                context
+            } else {
+                getActivity(context.baseContext)
+            }
+        }
+        return null
+    }
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    override fun onReceivedClientCertRequest(view: WebView, request: ClientCertRequest) {
+        // Let the WebViewActivity know the endpoint requires TLS Client Auth
+        isTLSClientAuthNeeded = true
+
+        // Aim to obtain the private key for the whole lifecycle of the WebViewActivity
+        val activity = getActivity(view.context)
+        if (activity == null || activity !is AppCompatActivity) return
+
+        activity.lifecycleScope.launch {
```

Comment:
we don't call anything `suspend`ing here, so `activity.lifecycleScope.launch { ... }` could be removed.